internal/fields/mappings.go

```diff
@@ -0,0 +1,577 @@
+	if diff := cmp.Diff(actualMappings, previewMappings, transformJSON); diff == "" {
+		logger.Debugf("No changes found in mappings")
+		return errs.Unique()
+	}
+
+	var rawPreview mappingDefinitions
+	err = json.Unmarshal(previewMappings, &rawPreview)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to unmarshal preview mappings (index template %s): %w", v.indexTemplateName, err))
+		return errs.Unique()
+	}
+	var rawActual mappingDefinitions
+	err = json.Unmarshal(actualMappings, &rawActual)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to unmarshal actual mappings (data stream %s): %w", v.dataStreamName, err))
+		return errs.Unique()
+	}
+
+	mappingErrs := compareMappings("", rawPreview, rawActual, v.Schema, v.LocalSchema)
+	errs = append(errs, mappingErrs...)
+
+	if len(errs) > 0 {
+		return errs.Unique()
+	}
+
+	return nil
+}
+
+func currentMappingPath(path, key string) string {
+	if path == "" {
+		return key
+	}
+	return fmt.Sprintf("%s.%s", path, key)
+}
+
+func mappingParameter(field string, definition mappingDefinitions) string {
+	fieldValue, ok := definition[field]
+	if !ok {
+		return ""
+	}
+	value, ok := fieldValue.(string)
+	if !ok {
+		return ""
+	}
+	return value
+}
+
+func isLocalFieldTypeArray(field string, schema []FieldDefinition) bool {
+	definition := findElementDefinitionForRoot("", field, schema)
+	if definition == nil {
+		return false
+	}
+	return definition.Type == "array"
+}
+
+func isEmptyObject(definition mappingDefinitions) bool {
+	// Example:
+	//  "_tmp": {
+	//    "type": "object"
+	//  },
+	if len(definition) != 1 {
+		return false
+	}
+	return mappingParameter("type", definition) == "object"
+}
+
+func isObject(definition mappingDefinitions) bool {
+	// Example:
+	// "http": {
+	//   "properties": {
+	// 	   "request": {
+	// 	     "properties": {
+	//         "method": {
+	//           "type": "keyword",
+	//           "ignore_above": 1024
+	//         }
+	//       }
+	//     }
+	//   }
+	// }
+	if len(definition) > 2 {
+		// there could also be a "dynamic: true" in the mapping
+		return false
+	}
+	field, ok := definition["properties"]
+	if !ok {
+		return false
+	}
+	if _, ok = field.(map[string]any); !ok {
+		return false
+	}
+	return true
+}
+
+func isObjectDynamic(definition mappingDefinitions) bool {
+	// Example:
+	//  "labels": {
+	//    "type": "object",
+	//    "dynamic": "true"
+	//  },
+	if len(definition) != 2 {
+		return false
+	}
+	fieldType := mappingParameter("type", definition)
+	fieldDynamic := mappingParameter("dynamic", definition)
+
+	if fieldType != "object" {
+		return false
+	}
+	if fieldDynamic != "true" {
+		return false
+	}
+	return true
+}
+
+func isMultiFields(definition mappingDefinitions) bool {
+	// Example:
+	//  "path": {
+	//    "type": "keyword",
+	//    "fields": {
+	//      "text": {
+	//        "type": "match_only_text"
+	//      }
+	//    }
+	//  },
+	_, ok := definition["type"]
+	if !ok {
+		return false
+	}
```

Comment:
I wanted to ensure that `fields` is not a regular field of the object instead of being part of the "multi-fields".
If it exists "type", I'd assume that this is an object definition.

But I guess, it could happen the same for "type". Would it be needed to check that the type of "type" is an string too ?

Reply:
Well, at the moment this function is called in `compareMappings` should not be an object (it should not have the `properties` field). So it should mean that it's already a field mapping (leaf node).

Reply:
Updated this function to use `mappingParameters` instead to get the value of "type".